scalems/wrappers/__init__.py

```diff
@@ -16,32 +15,109 @@
 
 Core components:
     make_input(**kwargs) -> SimulationInput:
+
         A simulation tool module provides a *make_input* command that accepts
         implementation-dependent named arguments and returns an object that may
         be used as simulation input within the context of the other module tools.
 
+        Proposal: At an abstract level, this contains everything needed to define the simulation 
```

Comment:
As noted below, I think "everything needed to define the simulation mathematically" is pretty undesirable.  Algorithms where the individual simulation is modified at the ensemble level don't want to do this.  Instead, make_input should contain everything required *internally* for the simulation.

Reply:
Yeah, I think we need better definitions.  

Maybe we need to come back to the simulation execution from files vs. modified in flight distinction.  For everything except gmxapi, the way you modify at an ensemble level is wait until it ends (or kill it, we can keep that option open), and restart the simulation.  The goal, as I remember to first support this mode, within a framework that also allows modification in flight afterwards. So from the external perspective, I would argue (and I think we agreed on this a while back?) it should probably look the same regardless of which mode was being used - just be really poorly performant if you try to do something frequently when with discrete simulations when that execution pattern is much better suited for in-flight changes.

I do think you need to define what "required internally" means.  I can't actually visualize what this means specifically. Is this the simulation state (i.e. the "checkpoint" abstractly defined)?

Reply:
The goal is to support the not-in-flight mode more generally without killing the specific support for in-flight abilities that we currently have in particular implementations.

I think this is really a danger of over-specifying.  A simple powerful interface is to say the input encapsulates everything the simulation needs internally and says what can be read or modified externally.  I would argue that is the appropriate level of encapsulation.  It doesn't require the external elements to be specified--that could involve the whole ensemble algorithm, so specifying it on a per simulation basis is impossible.

So I think a more relaxed specification here solves all the problems and doesn't create new ones.

Reply:
@peterkasson can you define "externally" and "internally" since these are not really terms that are standard.  This will help me be able to respond in general much more clearly.